# Deep Learning
## Assignment 2

Previously in 1_notmnist.ipynb, we created a pickle with formatted datasets for training, development and testing on the notMNIST dataset.

The goal of this assignment is to progressively train deeper and more accurate models using TensorFlow.


In [1]:
# These are all the modules we'll be using later. Make sure you can import them
# before proceeding further.
from __future__ import print_function
import numpy as np
import tensorflow as tf
from six.moves import cPickle as pickle
from six.moves import range

First reload the data we generated in 1_notmnist.ipynb.

In [2]:
pickle_file = 'notMNIST.pickle'

with open(pickle_file, 'rb') as f:
  save = pickle.load(f)
  train_dataset = save['train_dataset']
  train_labels = save['train_labels']
  valid_dataset = save['valid_dataset']
  valid_labels = save['valid_labels']
  test_dataset = save['test_dataset']
  test_labels = save['test_labels']
  del save  # hint to help gc free up memory
  print('Training set', train_dataset.shape, train_labels.shape)
  print('Validation set', valid_dataset.shape, valid_labels.shape)
  print('Test set', test_dataset.shape, test_labels.shape)

Training set (200000, 28, 28) (200000,)
Validation set (10000, 28, 28) (10000,)
Test set (10000, 28, 28) (10000,)


Reformat into a shape that's more adapted to the models we're going to train:

* data as a flat matrix,
* labels as float 1-hot encodings.

In [3]:
image_size = 28
num_labels = 10

def reformat(dataset, labels):
  dataset = dataset.reshape((-1, image_size * image_size)).astype(np.float32)
  # Map 0 to [1.0, 0.0, 0.0 ...], 1 to [0.0, 1.0, 0.0 ...]
  labels = (np.arange(num_labels) == labels[:,None]).astype(np.float32)
  return dataset, labels
train_dataset, train_labels = reformat(train_dataset, train_labels)
valid_dataset, valid_labels = reformat(valid_dataset, valid_labels)
test_dataset, test_labels = reformat(test_dataset, test_labels)
print('Training set', train_dataset.shape, train_labels.shape)
print('Validation set', valid_dataset.shape, valid_labels.shape)
print('Test set', test_dataset.shape, test_labels.shape)

Training set (200000, 784) (200000, 10)
Validation set (10000, 784) (10000, 10)
Test set (10000, 784) (10000, 10)


We're first going to train a multinomial logistic regression using simple gradient descent.

TensorFlow works like this:

* First you describe the computation that you want to see performed: what the inputs, the variables, and the operations look like. These get created as nodes over a computation graph. This description is all contained within the block below:

    with graph.as_default():
        ...

* Then you can run the operations on this graph as many times as you want by calling session.run(), providing it outputs to fetch from the graph that get returned. This runtime operation is all contained in the block below:

    with tf.Session(graph=graph) as session:
        ...

Let's load all the data into TensorFlow and build the computation graph corresponding to our training:


In [4]:
# With gradient descent training, even this much data is prohibitive.
# Subset the training data for faster turnaround.
train_subset = 10000

graph = tf.Graph()
with graph.as_default():

  # Input data.
  # Load the training, validation and test data into constants that are
  # attached to the graph.
  tf_train_dataset = tf.constant(train_dataset[:train_subset, :])
  tf_train_labels = tf.constant(train_labels[:train_subset])
  tf_valid_dataset = tf.constant(valid_dataset)
  tf_test_dataset = tf.constant(test_dataset)
  
  # Variables.
  # These are the parameters that we are going to be training. The weight
  # matrix will be initialized using random values following a (truncated)
  # normal distribution. The biases get initialized to zero.
  weights = tf.Variable(
    tf.truncated_normal([image_size * image_size, num_labels]))
  biases = tf.Variable(tf.zeros([num_labels]))
  
  # Training computation.
  # We multiply the inputs with the weight matrix, and add biases. We compute
  # the softmax and cross-entropy (it's one operation in TensorFlow, because
  # it's very common, and it can be optimized). We take the average of this
  # cross-entropy across all training examples: that's our loss.
  logits = tf.matmul(tf_train_dataset, weights) + biases
  loss = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(labels=tf_train_labels, logits=logits))
  
  # Optimizer.
  # We are going to find the minimum of this loss using gradient descent.
  optimizer = tf.train.GradientDescentOptimizer(0.5).minimize(loss)
  
  # Predictions for the training, validation, and test data.
  # These are not part of training, but merely here so that we can report
  # accuracy figures as we train.
  train_prediction = tf.nn.softmax(logits)
  valid_prediction = tf.nn.softmax(
    tf.matmul(tf_valid_dataset, weights) + biases)
  test_prediction = tf.nn.softmax(tf.matmul(tf_test_dataset, weights) + biases)

Let's run this computation and iterate:

In [5]:
num_steps = 801

def accuracy(predictions, labels):
  return (100.0 * np.sum(np.argmax(predictions, 1) == np.argmax(labels, 1))
          / predictions.shape[0])

with tf.Session(graph=graph) as session:
  # This is a one-time operation which ensures the parameters get initialized as
  # we described in the graph: random weights for the matrix, zeros for the
  # biases. 
  tf.global_variables_initializer().run()
  print('Initialized')
  for step in range(num_steps):
    # Run the computations. We tell .run() that we want to run the optimizer,
    # and get the loss value and the training predictions returned as numpy
    # arrays.
    _, l, predictions = session.run([optimizer, loss, train_prediction])
    if (step % 100 == 0):
      print('Loss at step %d: %f' % (step, l))
      print('Training accuracy: %.1f%%' % accuracy(
        predictions, train_labels[:train_subset, :]))
      # Calling .eval() on valid_prediction is basically like calling run(), but
      # just to get that one numpy array. Note that it recomputes all its graph
      # dependencies.
      print('Validation accuracy: %.1f%%' % accuracy(
        valid_prediction.eval(), valid_labels))
  print('Test accuracy: %.1f%%' % accuracy(test_prediction.eval(), test_labels))

Initialized
Loss at step 0: 17.330164
Training accuracy: 8.5%
Validation accuracy: 11.4%
Loss at step 100: 2.335239
Training accuracy: 71.1%
Validation accuracy: 69.8%
Loss at step 200: 1.866457
Training accuracy: 74.4%
Validation accuracy: 72.3%
Loss at step 300: 1.612761
Training accuracy: 76.1%
Validation accuracy: 73.1%
Loss at step 400: 1.441862
Training accuracy: 77.0%
Validation accuracy: 73.7%
Loss at step 500: 1.316399
Training accuracy: 77.8%
Validation accuracy: 74.1%
Loss at step 600: 1.219152
Training accuracy: 78.5%
Validation accuracy: 74.2%
Loss at step 700: 1.140921
Training accuracy: 79.1%
Validation accuracy: 74.4%
Loss at step 800: 1.076302
Training accuracy: 79.6%
Validation accuracy: 74.5%
Test accuracy: 82.3%


Let's now switch to stochastic gradient descent training instead, which is much faster.

The graph will be similar, except that instead of holding all the training data into a constant node, we create a Placeholder node which will be fed actual data at every call of session.run().


In [6]:
batch_size = 128

graph = tf.Graph()
with graph.as_default():

  # Input data. For the training data, we use a placeholder that will be fed
  # at run time with a training minibatch.
  tf_train_dataset = tf.placeholder(tf.float32,
                                    shape=(batch_size, image_size * image_size))
  tf_train_labels = tf.placeholder(tf.float32, shape=(batch_size, num_labels))
  tf_valid_dataset = tf.constant(valid_dataset)
  tf_test_dataset = tf.constant(test_dataset)
  
  # Variables.
  weights = tf.Variable(
    tf.truncated_normal([image_size * image_size, num_labels]))
  biases = tf.Variable(tf.zeros([num_labels]))
  
  # Training computation.
  logits = tf.matmul(tf_train_dataset, weights) + biases
  loss = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(labels=tf_train_labels, logits=logits))
  
  # Optimizer.
  optimizer = tf.train.GradientDescentOptimizer(0.5).minimize(loss)
  
  # Predictions for the training, validation, and test data.
  train_prediction = tf.nn.softmax(logits)
  valid_prediction = tf.nn.softmax(
    tf.matmul(tf_valid_dataset, weights) + biases)
  test_prediction = tf.nn.softmax(tf.matmul(tf_test_dataset, weights) + biases)

Let's run it:

In [7]:
num_steps = 3001

with tf.Session(graph=graph) as session:
  tf.global_variables_initializer().run()
  print("Initialized")
  for step in range(num_steps):
    # Pick an offset within the training data, which has been randomized.
    # Note: we could use better randomization across epochs.
    offset = (step * batch_size) % (train_labels.shape[0] - batch_size)
    # Generate a minibatch.
    batch_data = train_dataset[offset:(offset + batch_size), :]
    batch_labels = train_labels[offset:(offset + batch_size), :]
    # Prepare a dictionary telling the session where to feed the minibatch.
    # The key of the dictionary is the placeholder node of the graph to be fed,
    # and the value is the numpy array to feed to it.
    feed_dict = {tf_train_dataset : batch_data, tf_train_labels : batch_labels}
    _, l, predictions = session.run(
      [optimizer, loss, train_prediction], feed_dict=feed_dict)
    if (step % 500 == 0):
      print("Minibatch loss at step %d: %f" % (step, l))
      print("Minibatch accuracy: %.1f%%" % accuracy(predictions, batch_labels))
      print("Validation accuracy: %.1f%%" % accuracy(
        valid_prediction.eval(), valid_labels))
  print("Test accuracy: %.1f%%" % accuracy(test_prediction.eval(), test_labels))

Initialized
Minibatch loss at step 0: 20.277325
Minibatch accuracy: 10.9%
Validation accuracy: 10.1%
Minibatch loss at step 500: 1.413943
Minibatch accuracy: 77.3%
Validation accuracy: 75.4%
Minibatch loss at step 1000: 1.483913
Minibatch accuracy: 78.1%
Validation accuracy: 76.6%
Minibatch loss at step 1500: 0.922920
Minibatch accuracy: 81.2%
Validation accuracy: 77.2%
Minibatch loss at step 2000: 0.865631
Minibatch accuracy: 84.4%
Validation accuracy: 77.4%
Minibatch loss at step 2500: 1.032764
Minibatch accuracy: 78.9%
Validation accuracy: 77.8%
Minibatch loss at step 3000: 0.896759
Minibatch accuracy: 79.7%
Validation accuracy: 78.5%
Test accuracy: 86.5%


### Problem

Turn the logistic regression example with SGD into a 1-hidden layer neural network with rectified linear units nn.relu() and 1024 hidden nodes. This model should improve your validation / test accuracy.

### One Hidden Layer

In [8]:
num_nodes = 1024
batch_size = 128

graph = tf.Graph()
with graph.as_default():

  # Input data. For the training data, we use a placeholder that will be fed
  # at run time with a training minibatch.
  tf_train_dataset = tf.placeholder(tf.float32,
                                    shape=(batch_size, image_size * image_size))
  tf_train_labels = tf.placeholder(tf.float32, shape=(batch_size, num_labels))
  tf_valid_dataset = tf.constant(valid_dataset)
  tf_test_dataset = tf.constant(test_dataset)
  
  # Variables.
  weights_1 = tf.Variable(
    tf.truncated_normal([image_size * image_size, num_nodes]))
  biases_1 = tf.Variable(tf.zeros([num_nodes]))
  weights_2 = tf.Variable(
    tf.truncated_normal([num_nodes, num_labels]))
  biases_2 = tf.Variable(tf.zeros([num_labels]))
  
  # Training computation.
  relu_layer = tf.nn.relu(tf.matmul(tf_train_dataset, weights_1) + biases_1)
  logits = tf.matmul(relu_layer, weights_2) + biases_2
  loss = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(labels=tf_train_labels, logits=logits))
  
  # Optimizer.
  optimizer = tf.train.GradientDescentOptimizer(0.5).minimize(loss)
  
  # Predictions for the training, validation, and test data.
  train_prediction = tf.nn.softmax(logits)
  valid_prediction = tf.nn.softmax(
    tf.matmul(tf.nn.relu(tf.matmul(tf_valid_dataset, weights_1) + biases_1), weights_2) + biases_2)
  test_prediction = tf.nn.softmax(
      tf.matmul(tf.nn.relu(tf.matmul(tf_test_dataset, weights_1) + biases_1), weights_2) + biases_2)

In [10]:
num_steps = 10000

with tf.Session(graph=graph) as session:
  tf.global_variables_initializer().run()
  print("Initialized")
  for step in range(num_steps):
    # Pick an offset within the training data, which has been randomized.
    # Note: we could use better randomization across epochs.
    offset = (step * batch_size) % (train_labels.shape[0] - batch_size)
    # Generate a minibatch.
    batch_data = train_dataset[offset:(offset + batch_size), :]
    batch_labels = train_labels[offset:(offset + batch_size), :]
    # Prepare a dictionary telling the session where to feed the minibatch.
    # The key of the dictionary is the placeholder node of the graph to be fed,
    # and the value is the numpy array to feed to it.
    feed_dict = {tf_train_dataset : batch_data, tf_train_labels : batch_labels}
    _, l, predictions = session.run(
      [optimizer, loss, train_prediction], feed_dict=feed_dict)
    if (step % 500 == 0):
      print("Minibatch loss at step %d: %f" % (step, l))
      print("Minibatch accuracy: %.1f%%" % accuracy(predictions, batch_labels))
      print("Validation accuracy: %.1f%%" % accuracy(
        valid_prediction.eval(), valid_labels))
  print("Test accuracy: %.1f%%" % accuracy(test_prediction.eval(), test_labels))

Initialized
Minibatch loss at step 0: 334.516541
Minibatch accuracy: 9.4%
Validation accuracy: 20.2%
Minibatch loss at step 500: 22.166805
Minibatch accuracy: 82.8%
Validation accuracy: 80.3%
Minibatch loss at step 1000: 9.094169
Minibatch accuracy: 81.2%
Validation accuracy: 80.2%
Minibatch loss at step 1500: 5.162358
Minibatch accuracy: 87.5%
Validation accuracy: 80.5%
Minibatch loss at step 2000: 3.172457
Minibatch accuracy: 86.7%
Validation accuracy: 81.8%
Minibatch loss at step 2500: 3.100664
Minibatch accuracy: 85.9%
Validation accuracy: 82.3%
Minibatch loss at step 3000: 1.754658
Minibatch accuracy: 85.2%
Validation accuracy: 82.9%
Minibatch loss at step 3500: 3.197905
Minibatch accuracy: 85.2%
Validation accuracy: 82.1%
Minibatch loss at step 4000: 3.788253
Minibatch accuracy: 87.5%
Validation accuracy: 82.4%
Minibatch loss at step 4500: 4.550016
Minibatch accuracy: 84.4%
Validation accuracy: 81.0%
Minibatch loss at step 5000: 2.232993
Minibatch accuracy: 88.3%
Validation accur

### Two Hidden Layers using MomentumOptimizer

In [58]:
hidden_1 = 1024
hidden_2 = 500
batch_size = 128

import math

graph = tf.Graph()
with graph.as_default():

  # Input data. For the training data, we use a placeholder that will be fed
  # at run time with a training minibatch.
  tf_train_dataset = tf.placeholder(tf.float32,
                                    shape=(batch_size, image_size * image_size))
  tf_train_labels = tf.placeholder(tf.float32, shape=(batch_size, num_labels))
  tf_valid_dataset = tf.constant(valid_dataset)
  tf_test_dataset = tf.constant(test_dataset)
  
  # Variables.
  # Adjusted the stddev value as per https://discussions.udacity.com/t/assignment-2-2-hidden-layers-error/183933/4 
  # Note: any stddev that doesn't produce NAN is fine.
  weights_1 = tf.Variable(
    tf.truncated_normal([image_size * image_size, hidden_1], 
                        stddev=3.0 / math.sqrt(float(image_size * image_size + hidden_1))))
  biases_1 = tf.Variable(tf.zeros([hidden_1]))
  weights_2 = tf.Variable(
    tf.truncated_normal([hidden_1, hidden_2], 
                        stddev=3.0 / math.sqrt(float(hidden_1 + hidden_2))))
  biases_2 = tf.Variable(tf.zeros([hidden_2]))
  weights_3 = tf.Variable(
    tf.truncated_normal([hidden_2, num_labels], 
                        stddev=3.0 / math.sqrt(float(hidden_2 + num_labels))))
  biases_3 = tf.Variable(tf.zeros([num_labels]))
  
  # Training computation.
  relu_layer_1 = tf.nn.relu(tf.matmul(tf_train_dataset, weights_1) + biases_1)
  relu_layer_2 = tf.nn.relu(tf.matmul(relu_layer_1, weights_2) + biases_2)
  logits = tf.matmul(relu_layer_2, weights_3) + biases_3
  loss = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(labels=tf_train_labels, logits=logits))
  
  # Optimizer.
  # optimizer = tf.train.GradientDescentOptimizer(0.5).minimize(loss)
  optimizer = tf.train.MomentumOptimizer(0.05, 0.005).minimize(loss)
  
  # Predictions for the training, validation, and test data.
  train_prediction = tf.nn.softmax(logits)
  valid_prediction = tf.nn.softmax(      
    tf.matmul(tf.nn.relu(tf.matmul(tf.nn.relu(
        tf.matmul(tf_valid_dataset, weights_1) + biases_1), weights_2) + biases_2), weights_3) + biases_3)
  test_prediction = tf.nn.softmax(
      tf.matmul(tf.nn.relu(tf.matmul(tf.nn.relu(
        tf.matmul(tf_test_dataset, weights_1) + biases_1), weights_2) + biases_2), weights_3) + biases_3)

In [60]:
num_steps = 15001

with tf.Session(graph=graph) as session:
  tf.global_variables_initializer().run()
  print("Initialized")
  for step in range(num_steps):
    # Pick an offset within the training data, which has been randomized.
    # Note: we could use better randomization across epochs.
    offset = (step * batch_size) % (train_labels.shape[0] - batch_size)
    # Generate a minibatch.
    batch_data = train_dataset[offset:(offset + batch_size), :]
    batch_labels = train_labels[offset:(offset + batch_size), :]
    # Prepare a dictionary telling the session where to feed the minibatch.
    # The key of the dictionary is the placeholder node of the graph to be fed,
    # and the value is the numpy array to feed to it.
    feed_dict = {tf_train_dataset : batch_data, tf_train_labels : batch_labels}
    _, l, predictions = session.run(
      [optimizer, loss, train_prediction], feed_dict=feed_dict)
    if (step % 500 == 0):     
      print("Minibatch loss at step %d: %f" % (step, l))
      print("Minibatch accuracy: %.1f%%" % accuracy(predictions, batch_labels))
      print("Validation accuracy: %.1f%%" % accuracy(
        valid_prediction.eval(), valid_labels))
  print("Test accuracy: %.1f%%" % accuracy(test_prediction.eval(), test_labels))

Initialized
Minibatch loss at step 0: 3.878398
Minibatch accuracy: 10.2%
Validation accuracy: 22.2%
Minibatch loss at step 500: 0.411732
Minibatch accuracy: 85.9%
Validation accuracy: 84.2%
Minibatch loss at step 1000: 0.551803
Minibatch accuracy: 83.6%
Validation accuracy: 85.5%
Minibatch loss at step 1500: 0.257058
Minibatch accuracy: 93.0%
Validation accuracy: 86.6%
Minibatch loss at step 2000: 0.250747
Minibatch accuracy: 94.5%
Validation accuracy: 86.9%
Minibatch loss at step 2500: 0.396349
Minibatch accuracy: 89.8%
Validation accuracy: 87.3%
Minibatch loss at step 3000: 0.398312
Minibatch accuracy: 86.7%
Validation accuracy: 87.5%
Minibatch loss at step 3500: 0.384637
Minibatch accuracy: 88.3%
Validation accuracy: 88.1%
Minibatch loss at step 4000: 0.258426
Minibatch accuracy: 93.0%
Validation accuracy: 88.3%
Minibatch loss at step 4500: 0.284676
Minibatch accuracy: 90.6%
Validation accuracy: 88.2%
Minibatch loss at step 5000: 0.318603
Minibatch accuracy: 92.2%
Validation accurac

Result with GradientDescentOptimizer: 

##### Learning rate = 0.5
Minibatch loss at step 5000: 0.285956
Minibatch accuracy: 93.8%
Validation accuracy: 89.3%
Test accuracy: 94.8%


Result with MomentumOptimizer: 

##### learning rate = 0.5 , momentum = 0.5
Minibatch loss at step 5000: 0.347296
Minibatch accuracy: 89.1%
Validation accuracy: 88.5%
Test accuracy: 94.4%

##### learning rate = 0.5 , momentum = 0.05
Minibatch loss at step 5000: 0.286896
Minibatch accuracy: 93.8%
Validation accuracy: 89.2%
Test accuracy: 94.7%

##### learning rate = 0.5 , momentum = 0.005
Minibatch loss at step 5000: 0.290858
Minibatch accuracy: 91.4%
Validation accuracy: 88.2%
Test accuracy: 93.2%

##### learning rate = 0.1 , momentum = 0.005
Minibatch loss at step 10000: 0.175212
Minibatch accuracy: 95.3%
Validation accuracy: 89.4%
Test accuracy: 95.2%

##### learning rate = 0.05 , momentum = 0.005
Minibatch loss at step 15000: 0.195291
Minibatch accuracy: 94.5%
Validation accuracy: 89.7%
Test accuracy: 95.4%